### PR TITLE
[WIP] - Kubeadm phase ga

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR prototypes the approach described in [KEP kubeadm phases to beta](https://github.com/kubernetes/community/pull/1939).

As a result you can execute `kubeadm init [flags]` as today - for executing the entire init workflow -, or use `kubeadm init [phases] [flags]` for executing only single steps of the workflow. e.g.

```bash
    # to execute only certs phases
    kubeadm init certs [flags]
    
    # to execute only certs/apiserver phases
    kubeadm init certs/apiserver [flags]
```

**Which issue(s) this PR fixes**:
Fixes #https://github.com/kubernetes/kubeadm/issues/454

**Special notes for your reviewer**:
/CC @kubernetes/sig-cluster-lifecycle-pr-reviews 

The following commits contains the relevant changes:

* phases to beta
* cmd factories
* phased command

Other commits contains deleted files (many!) and autogenerate files.

The kubeadm init help shows all the available phases:
```
  preflight                  Run master pre-flight checks
  certs|certificates         Generates all PKI assets necessary to establish the control plane
    /ca                        Generates a self-signed kubernetes CA to provision identities for components of the cluster
    /apiserver                 Generates an API server serving certificate and key
    /apiserver-kubelet-client  Generates a client certificate for the API server to connect to the kubelets securely
    /etcd-ca                   Generates a self-signed CA to provision identities for etcd
    /etcd-server               Generates an etcd serving certificate and key
    /etcd-peer                 Generates an etcd peer certificate and key
    /etcd-healthcheck-client   Generates a client certificate for liveness probes to healthcheck etcd
    /apiserver-etcd-client     Generates a client certificate for the API server to connect to etcd securely
    /sa                        Generates a private key for signing service account tokens along with its public key
    /front-proxy-ca            Generates a front proxy CA certificate and key for a Kubernetes cluster
    /front-proxy-client        Generates a front proxy CA client certificate and key for a Kubernetes cluster
  kubeconfig                 Generates all kubeconfig files necessary to establish the control plane and the admin kubeconfig file
    /admin                     Generates a kubeconfig file for the admin to use and for kubeadm itself
    /kubelet                   Generates a kubeconfig file for the kubelet to use. Please note that this should be used *only* for bootstrapping purposes.
    /controller-manager        Generates a kubeconfig file for the controller manager to use
    /scheduler                 Generates a kubeconfig file for the scheduler to use
  audit-policy               Generates the audit-policy to use in the API server configuration (featureGated)
  controlplane               Generates all static Pod manifest files necessary to establish the control plane
    /apiserver                 Generates the API server static Pod manifest.
    /controller-manager        Generates the controller-manager static Pod manifest.
    /scheduler                 Generates the scheduler static Pod manifest.
  etcd                       Generates the static Pod manifest file for a local, single-node etcd instance
  init-kubeletconfig         (featureGated) Writes base configuration of kubelets to disk
  upload-kubeletconfig       (featureGated) Uploads kubelet config to a configMap and use it as a configSource for this node
  upload-config              Uploads the currently used configuration for kubeadm to a ConfigMap
  mark-master                Marks a node as master
  bootstrap-token            Makes all the bootstrap token configurations and creates an initial token
    /token                     Creates an initial bootstrap token to be used for node joining
    /cluster-info|clusterinfo  Uploads the cluster-info ConfigMap from the given kubeconfig file
    /allow-post-csr            Configures RBAC to allow node bootstrap tokens to post CSR in order for nodes to get long term certificate credentials
    /allow-auto-approve        Configures RBAC rules to allow the CSR approver controller automatically approve CSR from a node bootstrap token
  addons                     Installs required addons for passing Conformance tests
    /dns                       Installs the dns addon to a Kubernetes cluster
    /proxy                     Installs the proxy addon to a Kubernetes cluster
  selfhosting|self-hosting   Makes a kubeadm cluster self-hosted (featureGated)
```

**Release note**:
```release-note
TBD
```

### CODE REVIEW

**Comment by jamiehannaford** [comment]
@fabriziopandini Is there any way we can split up this PR into smaller chunks? 134 files is a lot to parse in one go! 😄 🤔

**Comment by fabriziopandini** [comment]
@jamiehannaford I'm going to remove temporarly auto-generated docs from the PR; this reduce the number of changes significantly, even if some check will complain.

 The following commits will contains the relevant changes:
- phases to beta
- cmd factories
- phased command
An this sum up to about 2000 lines of code for the full kubeadm init + phases, and then you will have some deleted files

File: cmd/kubeadm/app/cmd/util/phases/phases.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+	"github.com/spf13/cobra"
+)
+
+// PhaseWorkflow represents an order sequence of phases.
+type PhaseWorkflow []*Phase
+
+// HasByArgOrAlias returns true if the given phase (identified by phase.Arg)
+// is included in the workflow
+func (w PhaseWorkflow) HasByArgOrAlias(arg string) bool {
```
Comment (liztio) [inline cmd/kubeadm/app/cmd/util/phases/phases.go]: any reason this isn't `HasOrgOrAlias` like the individual phases?

File: cmd/kubeadm/app/cmd/util/phases/builder.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+func (pc *PhasedCommandBuilder) useLine() string {
+	if !strings.Contains(pc.Use, "[phases]") {
+		return fmt.Sprintf("%s [phases]", pc.Use)
+	}
+	return pc.Use
+}
+
+// descriptionWithPhases returns the PhasedCommand long description with the addition of the
+// message for using phases
+func (pc *PhasedCommandBuilder) descriptionWithPhases() string {
```
[the remark this answers is not in the document]
Comment (liztio) [reply cmd/kubeadm/app/cmd/util/phases/builder.go]: nevermind, found it in doc.go

File: cmd/kubeadm/app/cmd/util/factory/masterconfiguration.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		return dryRunDir, dryRunDir, dryRunDir, nil
+	}
+
+	return defaultPkiDir, constants.KubernetesDir, constants.GetStaticPodDirectory(), nil
+}
+
+// MasterConfiguration returns the MasterConfiguration instance.
+func (f *MasterConfigurationFactory) MasterConfiguration() *kubeadmapi.MasterConfiguration {
+	if f.masterConfigurationInstance == nil {
+		panic("Invalid operation. InitMasterConfiguration must be executed before GetMasterConfiguration")
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/cmd/util/factory/masterconfiguration.go]: Maybe we want to make the panic info more clear? I think users may not be sure what happened or what to do when they see this info.

Comment (xiangpengzhao) [reply cmd/kubeadm/app/cmd/util/factory/masterconfiguration.go]: same as below

**Review by jamiehannaford — COMMENTED** [review]
first round. I'll try to have a deeper look at this soon when I have time.

**Pull request closed without merging**